caffe2/operators/reduce_ops.cc

```diff
@@ -0,0 +1,146 @@
+/**
+ * Copyright (c) 2016-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "caffe2/operators/reduce_ops.h"
+
+namespace caffe2 {
+
+vector<TIndex> ConvertFromInputIndex(TIndex index, vector<TIndex>& dims) {
+  TIndex ndim = dims.size();
+  vector<TIndex> nd_idx(ndim);
+
+  for (TIndex i = ndim - 1; i >= 0 && index > 0; i--) {
+    nd_idx[i] = index % dims[i];
+    index /= dims[i];
+  }
+  return nd_idx;
+}
+
+TIndex ConvertToOutputIndex(
+    const vector<int>& axes,
+    const vector<TIndex>& nd_idx,
+    vector<TIndex>& dims) {
+  TIndex index = 0;
+  TIndex mul = 1;
+
+  for (TIndex i = dims.size() - 1, j = axes.size() - 1; i >= 0; i--) {
+    if (j >= 0 && axes[j] == i) {
+      j--;
+    } else {
+      index += nd_idx[i] * mul;
+      mul *= dims[i];
+    }
+  }
+  return index;
+}
+
+template <typename T, class Context>
+void ComputeSum(
```

Comment:
Let's make computeSum more general, so it can support max/min/log/... later you don't need to refactor the code